lib/Dialect/Arc/Transforms/GroupResetsAndEnables.cpp

```diff
@@ -0,0 +1,160 @@
+                                PatternRewriter &rewriter) const override {
+    // Group similar resets into single IfOps
+    // Create a list of reset values and map from them to the states they reset
+    llvm::MapVector<mlir::Value, SmallVector<scf::IfOp>> resetMap;
+
+    for (auto ifOp : clockTreeOp.getBody().getOps<scf::IfOp>())
+      if (ifOp.getResults().empty())
+        resetMap[ifOp.getCondition()].push_back(ifOp);
+
+    // TODO: Check that conflicting memory effects aren't being reordered
+
+    // Combine IfOps
+    bool changed = false;
+    for (auto &[cond, oldOps] : resetMap) {
+      if (oldOps.size() <= 1)
+        continue;
+      scf::IfOp lastIfOp = oldOps.pop_back_val();
+      for (auto thisOp : oldOps) {
+        // Inline the before and after region inside the original If
+        rewriter.eraseOp(thisOp.thenBlock()->getTerminator());
+        rewriter.inlineBlockBefore(thisOp.thenBlock(),
+                                   &lastIfOp.thenBlock()->front());
+        // Check we're not inlining an empty block
+        if (auto *elseBlock = thisOp.elseBlock()) {
+          rewriter.eraseOp(elseBlock->getTerminator());
+          if (auto *lastElseBlock = lastIfOp.elseBlock())
+            rewriter.inlineBlockBefore(elseBlock,
+                                       &lastIfOp.elseBlock()->front());
+          else {
+            lastElseBlock = rewriter.createBlock(&lastIfOp.getElseRegion());
+            rewriter.setInsertionPointToEnd(lastElseBlock);
+            auto yieldOp =
+                rewriter.create<scf::YieldOp>(rewriter.getUnknownLoc());
+            rewriter.inlineBlockBefore(thisOp.elseBlock(), yieldOp);
+          }
+        }
+        rewriter.eraseOp(thisOp);
+        changed = true;
+      }
+    }
+    return success(changed);
+  }
+};
+
+struct EnableGroupingPattern : public OpRewritePattern<ClockTreeOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(ClockTreeOp clockTreeOp,
+                                PatternRewriter &rewriter) const override {
+    // Amass regions that we want to group enables in
+    SmallVector<Region *> groupingRegions;
+    groupingRegions.push_back(&clockTreeOp.getBody());
+    for (auto ifOp : clockTreeOp.getBody().getOps<scf::IfOp>()) {
+      groupingRegions.push_back(&ifOp.getThenRegion());
+      groupingRegions.push_back(&ifOp.getElseRegion());
+    }
+
+    bool changed = false;
+    for (auto *region : groupingRegions) {
+      llvm::MapVector<mlir::Value, SmallVector<StateWriteOp>> enableMap;
+      auto writeOps = region->getOps<StateWriteOp>();
+      for (auto writeOp : writeOps) {
+        if (writeOp.getCondition())
+          enableMap[writeOp.getCondition()].push_back(writeOp);
+      }
+      for (auto &[enable, writeOps] : enableMap) {
+        // Only group if multiple writes share an enable
+        if (writeOps.size() <= 1)
+          continue;
+        if (isa<scf::IfOp>(region->getParentOp()))
```

Comment:
Agreed! Hoped to do this originally but ran into some baffling errors - seems to work fine checking the trait now though, so not sure what I missed before!

Reply:
Oops, just realised I misread this - I updated it to check the NoTerminator trait on the parent op instead. Is checking the last op's trait preferable?

Reply:
I initially thought of the `Terminator` trait on the last op in the block, but that approach is also fine. The case where it could fail is when you don't erase the terminator before inlining (or moving ops over in another way) and thus the `NoTerminator` invariant will be invalid for the time being, but that isn't an issue in your implementation.